aten/src/ATen/native/mkl/SparseCsrLinearAlgebra.cpp

```diff
@@ -0,0 +1,217 @@
+    MKL_INT nrows,
+    MKL_INT ncols,
+    MKL_INT dense_ncols) {
+  sparse_matrix_t A = 0;
+  matrix_descr desc;
+  desc.type = SPARSE_MATRIX_TYPE_GENERAL;
+  int retval = mkl_sparse_s_create_csr(
+      &A,
+      SPARSE_INDEX_BASE_ZERO,
+      nrows,
+      ncols,
+      crow_indices,
+      crow_indices + 1,
+      col_indices,
+      values);
+  TORCH_CHECK(
+      retval == 0, "mkl_sparse_s_create_csr failed with error code: ", retval);
+
+  mkl_sparse_s_mm(
+      SPARSE_OPERATION_NON_TRANSPOSE,
+      alpha,
+      A,
+      desc,
+      SPARSE_LAYOUT_ROW_MAJOR,
+      dense,
+      dense_ncols,
+      dense_ncols,
+      beta,
+      res,
+      dense_ncols);
+  mkl_sparse_destroy(A);
+}
+
+static inline void sparse_mm_mkl_impl(
+    double* res,
+    MKL_INT* col_indices,
+    MKL_INT* crow_indices,
+    double* values,
+    double* dense,
+    double* t,
+    double alpha,
+    double beta,
+    MKL_INT nrows,
+    MKL_INT ncols,
+    MKL_INT dense_ncols) {
+  sparse_matrix_t A = 0;
+  matrix_descr desc;
+  desc.type = SPARSE_MATRIX_TYPE_GENERAL;
+  int retval = mkl_sparse_d_create_csr(
+      &A,
+      SPARSE_INDEX_BASE_ZERO,
+      nrows,
+      ncols,
+      crow_indices,
+      crow_indices + 1,
+      col_indices,
+      values);
+  TORCH_CHECK(
+      retval == 0, "mkl_sparse_d_create_csr failed with error code: ", retval);
+
+  mkl_sparse_d_mm(
+      SPARSE_OPERATION_NON_TRANSPOSE,
+      alpha,
+      A,
+      desc,
+      SPARSE_LAYOUT_ROW_MAJOR,
+      dense,
+      dense_ncols,
+      dense_ncols,
+      beta,
+      res,
+      dense_ncols);
+  mkl_sparse_destroy(A);
+}
+
+template <typename scalar_t>
+static inline void sparse_mm_mkl_template(
+    Tensor& res,
+    const Tensor& col_indices,
+    const Tensor& crow_indices,
+    const Tensor& values,
+    const Tensor& dense,
+    const Tensor& t,
+    const Scalar& alpha,
+    const Scalar& beta,
+    IntArrayRef size,
+    IntArrayRef dense_size) {
+  sparse_mm_mkl_impl(
+      res.data_ptr<scalar_t>(),
+      col_indices.data_ptr<MKL_INT>(),
+      crow_indices.data_ptr<MKL_INT>(),
+      values.data_ptr<scalar_t>(),
+      dense.data_ptr<scalar_t>(),
+      t.data_ptr<scalar_t>(),
+      alpha.to<scalar_t>(),
+      beta.to<scalar_t>(),
+      size[0],
+      size[1],
+      dense_size[1]);
+}
+
+static bool inline is_mkl_int32_index() {
+#ifdef MKL_ILP64
+  return false;
+#else
+  return true;
+#endif
+}
+
+Tensor& _sparse_mm_mkl_(
+    Tensor& self,
+    const SparseTensor& sparse_,
+    const Tensor& dense,
+    const Tensor& t,
+    const Scalar& alpha,
+    const Scalar& beta) {
+  if (is_mkl_int32_index()) {
+    if (sparse_.crow_indices().scalar_type() != kInt) {
+      TORCH_WARN(
+          "Pytorch is compiled with MKL LP64 and will convert crow_indices to int32.");
+    }
+    if (sparse_.col_indices().scalar_type() != kInt) {
+      TORCH_WARN(
+          "Pytorch is compiled with MKL LP64 and will convert col_indices to int32.");
+    }
```

Comment:
I'm not convinced we shouldn't just hard error in this situation and force the user to do the conversion themselves. (Speaking of conversions, is there a way to change the index dtype only?)

Reply:
I don't there is a way to change the index type only. We do this conversion within this code because only MKL functions need 32 bit int indexing and everything else (I think CUDA too) can work with 64 bit. If only a single function being called by the user requires MKL, I think it makes sense to do this transparently for MKL rather than enforcing 32 bit indexing for all operations.